src/subcommand/export.rs

```diff
@@ -0,0 +1,338 @@
+        break;
+      }
+    }
+    Ok(())
+  }
+
+  fn prepare_directory_tree(&self) -> Result {
+    fs::create_dir_all(self.numbers_dir())?;
+    Ok(())
+  }
+
+  fn numbers_dir(&self) -> PathBuf {
+    self.output_dir.join("numbers")
+  }
+
+  fn get_written_numbers(&self) -> Result<HashSet<u64>> {
+    Ok(
+      fs::read_dir(self.numbers_dir())?
+        .filter_map(|dir_entry| -> Option<u64> {
+          u64::from_str(dir_entry.ok()?.path().file_stem()?.to_str()?).ok()
+        })
+        .collect(),
+    )
+  }
+
+  fn export_inscription_by_id(
+    &self,
+    index: &Index,
+    id: InscriptionId,
+    entry: InscriptionEntry,
+  ) -> Result {
+    let inscription = index
+      .get_inscription_by_id(id)?
+      .ok_or_else(|| anyhow!("inscription not found: {id}"))?;
+    let number = entry.number;
+    match inscription.body() {
+      None => log::info!("inscription {number} has no body"),
+      Some(body) => {
+        let file = self.file_name(&inscription, number);
+        fs::write(file, body)?;
+      }
+    }
+    Ok(())
+  }
+
+  fn file_name(&self, inscription: &Inscription, number: u64) -> PathBuf {
+    let mut file = self.numbers_dir().join(number.to_string());
+    let extension = Export::file_extension(inscription, number);
+    if let Some(extension) = extension {
+      file.set_extension(extension);
+    }
+    file
+  }
+
+  fn file_extension(inscription: &Inscription, number: u64) -> Option<&'static str> {
+    match inscription.content_type() {
+      None => {
+        log::info!("content_type missing, writing inscription {number} without file extension");
+        None
+      }
+      Some(content_type) => match Media::extension_for_content_type(content_type) {
+        Err(message) => {
+          log::info!("{message}, writing inscription {number} without file extension");
+          None
+        }
+        Ok(extension) => Some(extension),
+      },
+    }
+  }
+
+  fn add_progress_bar<Iter, T>(iterator: Iter) -> Box<(dyn Iterator<Item = T>)>
+  where
+    Iter: Iterator<Item = T> + ExactSizeIterator + 'static,
+  {
+    if cfg!(test) || log_enabled!(log::Level::Info) || integration_test() {
+      Box::new(iterator)
+    } else {
+      Box::new(iterator.into_iter().progress_with_style(
+        ProgressStyle::with_template("[exporting inscriptions] {wide_bar} {pos}/{len}").unwrap(),
+      ))
+    }
+  }
+
+  fn retry<F, T>(message: &str, mut f: F) -> Result<T>
+  where
+    F: FnMut() -> Result<T>,
+  {
+    let mut tries = 5;
+    loop {
+      match f() {
+        Err(err) => {
+          if tries > 0 {
+            tries -= 1;
+            log::warn!("{message}: {err}");
+          } else {
+            return Err(err);
+          }
+        }
+        Ok(t) => return Ok(t),
+      }
+    }
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use {
+    super::*,
+    crate::{index::tests::Context, test::inscription},
+    pretty_assertions::assert_eq,
+  };
+
+  impl Context {
```

Comment:
These are additions to `index::test::Context` that are only relevant for `export` tests. Personally I like this structure, but I could see how someone might find this confusing. I'd be happy to move `Context` into a common test helper module.